Deprecate `Time.toString` methods

mihails-strasuns-sociomantic opened the pull request:
First step before removing them in ocean v4.0.0 and breaking one of import cycles vanilla druntime complains about.

mihails-strasuns-sociomantic committed:
Implement pretty printing time formatter

To be used as a replacement for own `Time.toString` method

mihails-strasuns-sociomantic commented:
Competing proposal to https://github.com/sociomantic-tsunami/ocean/pull/93

mihails-strasuns-sociomantic commented:
Reason why I would prefer such more intrusive approach should be preferred is that I view every single circular import as a technical debt. You do it because it is often fast and convenient, but in the long run it is always a problem that needs to be solved. Circular imports break modularization, making impossible to import only subset of library functionality you really use and often harming build times as a side effect.

While the matter is not important enough to hide and kill every circular import case intentionally, it should not be hidden with workarounds if such case is shown by compiler/runtime.

In this specific case I also don't think it was a good idea in the first place to do any kind of non-trivial time formatting inside default `toString` method, especially considering it relies on global state.

mathias-lang-sociomantic commented:
Noticed, and I think the concept is sane. I would like to give it a look on a day I'm a bit more functional (feel free to ping me from Wednesday on ;) ).
Ultimately I think we could/should bring both P.R. in, because those static ctors are also nasty. But we can keep that discussion on the other P.R.

mihails-strasuns-sociomantic commented:
Wednesday ping @mathias-lang-sociomantic :)

mathias-lang-sociomantic commented on src/ocean/text/convert/DateTime_tango.d:
```diff
@@ -970,6 +970,41 @@ private struct Result
+{
+    private Time value;
+
+    public void toString (void delegate(cstring) sink)
```
`FormatterSink` ?

mathias-lang-sociomantic commented on src/ocean/text/convert/DateTime_tango.d:
```diff
@@ -970,6 +970,41 @@ private struct Result
+    {
+        // Layout defaults to 'G'
+        scope dg = (cstring s) { sink(s); return s.length; };
+        DateTimeDefault.format(dg, this.value, "");
```
Hum since those two differ I think we can just change `format` here... Will look into it

mihails-strasuns-sociomantic replied:
What do you mean?

mathias-lang-sociomantic replied:
This uses a `size_t`-returning delegate and shouldn't. But we can take care of that later.

mathias-lang-sociomantic commented on src/ocean/text/convert/Formatter_test.d:
```diff
@@ -546,6 +546,22 @@ unittest
+        "{ ticks_: 621360288000000000 }"
+    );
+}
+
```
I think this belong to `Time_test` more than formatter's

mathias-lang-sociomantic commented on src/ocean/time/Time_test.d:
```diff
@@ -0,0 +1,57 @@
+    test (TimeSpan.fromMinutes(50).time.minutes is 50);
+    test (TimeSpan.fromMinutes(5000).time.minutes is 5000 % 60);
+    test (TimeSpan.fromHours(23).time.hours is 23);
+    test (TimeSpan.fromHours(5000).time.hours is 5000 % 24);
```
Nitpick: extra space before parent

mihails-strasuns-sociomantic committed:
Split Time tests into separate module

Helps to test more interesting things without introducing circular
dependencies

mihails-strasuns-sociomantic committed:
Deprecate `Time.toString` methods

mihails-strasuns-sociomantic commented:
Updated:
- moved all tests to `Time_test.d`
- whitespaces
- `FormatterSink`

mathias-lang-sociomantic commented on src/ocean/time/Time_test.d:
```diff
@@ -0,0 +1,84 @@
+    test(tod.hours is 1);
+    test(tod.minutes is 2);
+    test(tod.seconds is 3);
+    test(tod.millis is 4);
```
All those tests could be converted to `test!("...")` but it's a minor point.

mathias-lang-sociomantic reviewed (approved):
LGTM

nemanja-boric-sociomantic commented on src/ocean/time/Time_test.d:
```diff
@@ -51,3 +53,16 @@ unittest
+{
+    test!("==")(
+        format("{}", asPrettyStr(Time.epoch1970)),
+        "01/01/70 00:00:00"
```
I sure hope the locale on the test machines will not change :-)

mihails-strasuns-sociomantic replied:
A _lot_ of tests will break when that happens :)

mihails-strasuns-sociomantic commented:
Thanks!
